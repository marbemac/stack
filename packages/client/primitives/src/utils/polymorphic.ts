/* eslint-disable @typescript-eslint/ban-types */
// Adapted from https://github.com/fabien-ml/kobalte/blob/8ab09d87f7ea68f9fb79d32458a06dd3ae62029d/packages/utils/src/polymorphic.tsx
// asProps design came out of this discussion -> https://github.com/orgs/kobaltedev/discussions/138#discussioncomment-5245471

import type { SetRequired } from '@marbemac/utils-types';
import type { Component, ComponentProps as BaseComponentProps, JSX } from 'solid-js';

/** All HTML and SVG elements. */
export type DOMElements = keyof JSX.IntrinsicElements;

/** Any HTML element or SolidJS component. */
export type ElementType<Props = any> = DOMElements | Component<Props>;

/**
 * Allows for extending a set of props (`Source`) by an overriding set of props (`Override`),
 * ensuring that any duplicates are overridden by the overriding set of props.
 */
export type OverrideProps<Source = {}, Override = {}> = Omit<Source, keyof Override> & Override;

/** The `as` prop type. */
export type As<Props = any> = ElementType<Props>;

/** Clean-up autocomplete and omit a bunch of props we don't use */
export type ComponentProps<T extends As> = Omit<BaseComponentProps<T>, keyof IgnoredHTMLAttributes<T>>;

export type IgnoredHTMLAttributes<T> = Pick<
  JSX.HTMLAttributes<T>,
  // Feel free to remove any items from this list if any end up being needed
  | '$ServerOnly'
  | 'about'
  | 'accessKey'
  | 'aria-atomic'
  | 'aria-flowto'
  | 'autocapitalize'
  | 'class'
  | 'classList'
  | 'color'
  | 'contenteditable'
  | 'contextmenu'
  | 'datatype'
  | 'dir'
  | 'exportParts'
  | 'exportparts'
  | 'inert'
  | 'inlist'
  | 'inputMode'
  | 'inputmode'
  | 'itemId'
  | 'itemProp'
  | 'itemRef'
  | 'itemScope'
  | 'itemType'
  | 'itemid'
  | 'itemprop'
  | 'itemref'
  | 'itemscope'
  | 'itemtype'
  | 'onabort'
  | 'onanimationend'
  | 'onanimationiteration'
  | 'onanimationstart'
  | 'onauxclick'
  | 'onbeforeinput'
  | 'onblur'
  | 'oncanplay'
  | 'oncanplaythrough'
  | 'onchange'
  | 'onclick'
  | 'oncontextmenu'
  | 'ondblclick'
  | 'ondrag'
  | 'ondragend'
  | 'ondragenter'
  | 'ondragleave'
  | 'ondragover'
  | 'ondragstart'
  | 'ondrop'
  | 'ondurationchange'
  | 'onemptied'
  | 'onended'
  | 'onerror'
  | 'onfocus'
  | 'ongotpointercapture'
  | 'oninput'
  | 'oninvalid'
  | 'onkeydown'
  | 'onkeypress'
  | 'onkeyup'
  | 'onload'
  | 'onloadeddata'
  | 'onloadedmetadata'
  | 'onloadstart'
  | 'onlostpointercapture'
  | 'onmousedown'
  | 'onmouseenter'
  | 'onmouseleave'
  | 'onmousemove'
  | 'onmouseout'
  | 'onmouseover'
  | 'onmouseup'
  | 'onpause'
  | 'onplay'
  | 'onplaying'
  | 'onpointercancel'
  | 'onpointerdown'
  | 'onpointerenter'
  | 'onpointerleave'
  | 'onpointermove'
  | 'onpointerout'
  | 'onpointerover'
  | 'onpointerup'
  | 'onprogress'
  | 'onratechange'
  | 'onreset'
  | 'onscroll'
  | 'onseeked'
  | 'onseeking'
  | 'onselect'
  | 'onstalled'
  | 'onsubmit'
  | 'onsuspend'
  | 'ontimeupdate'
  | 'ontouchcancel'
  | 'ontouchend'
  | 'ontouchmove'
  | 'ontouchstart'
  | 'ontransitionstart'
  | 'ontransitionend'
  | 'ontransitionrun'
  | 'ontransitioncancel'
  | 'onvolumechange'
  | 'onwaiting'
  | 'onwheel'
  | 'oncompositionend'
  | 'oncompositionstart'
  | 'oncompositionupdate'
  | 'oncopy'
  | 'oncut'
  | 'ondragexit'
  | 'onencrypted'
  | 'onfocusout'
  | 'onfocusin'
  | 'onpaste'
  | 'part'
  | 'prefix'
  | 'property'
  | 'resource'
  | 'slot'
  | 'tabindex'
  | 'typeof'
  | 'vocab'
>;

/** Props object that includes the `as` prop. */
export type PolymorphicProps<Type extends As = As, Props = {}> = OverrideProps<
  ComponentProps<Type>,
  // Props & { as?: Type }
  Props & AsProp<Type>
>;

interface AsProp<T extends As, P = ComponentProps<T>> {
  as?: T;
  asProps?: P;
}

type WithAs<T extends As> = SetRequired<AsProp<T>, 'as'>;

/** A component with the `as` prop. */
export type PolymorphicComponent<DefaultType extends As, Props = {}> = {
  <Type extends As>(props: PolymorphicProps<Type, Props> & WithAs<Type>): JSX.Element;
  (props: PolymorphicProps<DefaultType, Props>): JSX.Element;
};

/**
 * Create a component with the type cast to `PolymorphicComponent`.
 * You have to use `Dynamic` internally and pass the `as` prop to handle polymorphism correctly.
 */
export function createPolymorphicComponent<DefaultType extends As, Props = {}>(
  component: Component<PolymorphicProps<DefaultType, Props>>,
) {
  return component as unknown as PolymorphicComponent<DefaultType, Props>;
}
