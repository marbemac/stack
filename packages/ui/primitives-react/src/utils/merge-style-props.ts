/**
 * Merge default and component props into a single props object.
 * @param defaultProps The default props, will be overridden by component props.
 * @param props The component `props` object.
 * @example
 * // mergedProps = defaultProps <== props
 */
export function mergeStyleProps<T extends Record<string, any>>(defaultProps: Partial<T>, props: T): T {
  return Object.assign({}, defaultProps, props);
}
