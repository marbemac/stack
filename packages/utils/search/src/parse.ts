import type { ILexingResult, IRecognitionException, Lexer } from 'chevrotain';

import { SearchLexer } from './grammar/lexer.ts';
import { SearchParser } from './grammar/parser.ts';
import type { SearchString } from './types.ts';

let lexerSingleton: Lexer;
let parserSingleton: SearchParser;

export interface ParseQueryOpts<T extends InputType> {
  input: SearchString;
  inputType: T;
  lexer?: Lexer;
  parser?: SearchParser;
}

type InputType = 'searchQuery' | 'selectClause' | 'selectExpression' | 'fromClause' | 'whereClause' | 'whereExpression';

export const parseSearch = <T extends InputType>({
  input,
  inputType,
  ...rest
}: ParseQueryOpts<T>): {
  cst: ReturnType<SearchParser[T]>;
  errors: IRecognitionException[];
  lexResult: ILexingResult;
} => {
  let lexer = rest.lexer || lexerSingleton;
  if (!lexer) {
    lexer = lexerSingleton = new SearchLexer();
  }

  const lexResult = lexer.tokenize(input);

  let parser = rest.parser || parserSingleton;
  if (!parser) {
    parser = parserSingleton = new SearchParser();
  }

  // "input" is a setter which will reset the parser's state.
  parser.input = lexResult.tokens;

  const cst = parser[inputType]() as ReturnType<SearchParser[T]>;

  return { cst, errors: parser.errors, lexResult };
};
