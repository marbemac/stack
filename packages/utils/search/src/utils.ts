import {
  type FunctionAstNode,
  type IsNegatableNode,
  type IsSortableNode,
  type QualifierAstNode,
  type SearchNode,
} from './visitor.ts';

export function stringifySearchAstNode(token?: SearchNode) {
  if (!token) return '';

  switch (token.type) {
    case 'search_query':
      return [
        stringifySearchAstNode(token.fromClause),
        stringifySearchAstNode(token.selectClause),
        stringifySearchAstNode(token.whereClause),
      ]
        .filter(Boolean)
        .join('\n');
    case 'from_clause':
      return `FROM ${token.table}`;
    case 'select_clause':
      return token.expr.columns.length ? `SELECT ` : '';
    case 'select_expr':
      return token.columns.length ? token.columns.map(stringifySearchAstNode).join(' ') : '';
    case 'where_clause':
      return token.expr.conditions.length ? `WHERE ` : '';
    case 'where_expr':
      return token.conditions.length ? token.conditions.map(stringifySearchAstNode).join(' ') : '';
    case 'qualifier':
      return handleNegation(token, stringifyQualifier(token));
    case 'function':
      return handleSortDir(token, handleNegation(token, stringifyFunction(token)));
    case 'function_arg':
      return token.vals.map(stringifySearchAstNode).join(' ');
    case 'qualifier_key':
      return token.value;
    case 'bracket_list':
      return `[${token.values.map(stringifySearchAstNode).join(', ')}]`;
    case 'relative_date':
      return `${token.sign}${token.value}${token.unit}`;
    case 'text':
      return handleSortDir(token, token.quoted ? `"${token.value}"` : token.value);
    case 'number':
      return token.value;
    case 'boolean':
      return token.value;
  }
}

function stringifyQualifier(token: QualifierAstNode) {
  let stringifiedToken = stringifySearchAstNode(token.lhs);

  if (token.rhs) {
    stringifiedToken += ':';
    stringifiedToken += token.op || '';
    stringifiedToken += stringifySearchAstNode(token.rhs);
  }

  return stringifiedToken;
}

function stringifyFunction(token: FunctionAstNode) {
  let stringifiedToken = `${token.name}(`;

  if (token.args) {
    stringifiedToken += token.args.map(stringifySearchAstNode).join(', ');
  }

  stringifiedToken += ')';

  if (token.rhs) {
    stringifiedToken += ':';
    stringifiedToken += token.op || '';
    stringifiedToken += stringifySearchAstNode(token.rhs);
  }

  return stringifiedToken;
}

function handleSortDir(token: IsSortableNode, stringifiedToken: string) {
  if (!token.sort) return stringifiedToken;
  return `${token.sort === 'asc' ? '+' : '-'}${stringifiedToken}`;
}

function handleNegation(token: IsNegatableNode, stringifiedToken: string) {
  return token.negated ? `!${stringifiedToken}` : stringifiedToken;
}
