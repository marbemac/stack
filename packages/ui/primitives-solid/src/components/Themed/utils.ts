import { isFunction } from '@kobalte/utils';
import { mergeProps } from 'solid-js';

import type { ComponentsConfig } from './components-config.ts';
import { useTheme } from './theme-context.ts';

function useComponentTheme<T extends keyof ComponentsConfig>(component: T) {
  return useTheme()?.components?.[component] ?? undefined;
}

/**
 * Resolve the `slotClasses` provided in the component theme configuration.
 */
export function useThemeClasses<Slots extends string>(component: keyof ComponentsConfig, props: any) {
  const classes = useComponentTheme(component)?.slotClasses ?? {};

  return (isFunction(classes) ? classes(props) : classes) as Partial<Record<Slots, string>>;
}

/**
 * Merge default, theme and component props into a single props object.
 * @param component The name of the component to look for in the theme.
 * @param defaultProps The default props, will be overridden by theme and component props.
 * @param props The component `props` object.
 * @example
 * // mergedProps = defaultProps <== themeProps <== props
 */
export function mergeThemeProps<T extends Record<string, any>>(
  component: keyof ComponentsConfig,
  defaultProps: Partial<T>,
  props: T,
): T {
  const themeProps = () => useComponentTheme(component)?.defaultProps ?? {};

  return mergeProps(defaultProps, themeProps, props);
}
