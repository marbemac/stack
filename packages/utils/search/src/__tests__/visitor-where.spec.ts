import { describe, expect, it } from 'vitest';

import { parseQuery } from '../parse.ts';
import type { SearchString } from '../types.ts';
import { treeTransformer } from '../utils.ts';
import { createSearchVisitor } from '../visitor.ts';
import { whereCaseGroups } from './fixtures/common.ts';

const prettyPrintParseResult = (input: string) => {
  const { cst } = parseQuery({ input: input as SearchString, inputType: 'whereExpression' });

  const visitor = createSearchVisitor();
  const ast = visitor.whereExpression(cst.children);

  /**
   * Just cleans up the snapshots a little bit for easier review
   */
  return treeTransformer({
    tree: ast,
    transform: token => {
      // @ts-expect-error remove `quoted` property if false, since that's the default
      if (token.quoted === false) delete token.quoted;

      // @ts-expect-error remove `negated` property if false, since that's the default
      if (token.negated === false) delete token.negated;

      return token;
    },
  });
};

describe.each(Object.entries(whereCaseGroups))('where %s', (_, cases) => {
  it.each(cases)('%s', (_, input) => {
    expect({
      input,
      result: prettyPrintParseResult(input),
    }).toMatchSnapshot();
  });
});

it('debug', { skip: true }, () => {
  const input = 'user:[jane, "john doe"] release:[12.0]';
  // const input = 'one:two "foo bar" three:four';
  // const input = 'last_seen:+1d';

  const lexResult = { tokens: [] };
  // const { lexResult } = parseQuery({ input: input as SearchString, inputType: 'whereExpression' });

  expect(lexResult.tokens).toMatchInlineSnapshot(`[]`);

  expect(prettyPrintParseResult(input)).toMatchInlineSnapshot();
});
