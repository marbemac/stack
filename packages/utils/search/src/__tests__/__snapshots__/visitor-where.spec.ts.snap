// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`where boolean > false 1`] = `
{
  "input": "plan:false",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "plan",
          },
          "op": "",
          "rhs": {
            "parsed": false,
            "type": "boolean",
            "value": "false",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where boolean > true 1`] = `
{
  "input": "plan:true",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "plan",
          },
          "op": "",
          "rhs": {
            "parsed": true,
            "type": "boolean",
            "value": "true",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where dates > date only 1`] = `
{
  "input": "last_seen:>2024-01-12",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "last_seen",
          },
          "op": ">",
          "rhs": {
            "type": "iso_8601_date",
            "value": "2024-01-12",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where dates > date time 1`] = `
{
  "input": "last_seen:2024-01-12T10:15:01",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "last_seen",
          },
          "op": "",
          "rhs": {
            "type": "iso_8601_date",
            "value": "2024-01-12T10:15:01",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where functions > boolean arguments 1`] = `
{
  "input": "my_func(true, false)",
  "result": {
    "ast": {
      "conditions": [
        {
          "args": [
            {
              "position": 0,
              "type": "function_arg",
              "vals": [
                {
                  "parsed": true,
                  "type": "boolean",
                  "value": "true",
                },
              ],
            },
            {
              "position": 1,
              "type": "function_arg",
              "vals": [
                {
                  "parsed": false,
                  "type": "boolean",
                  "value": "false",
                },
              ],
            },
          ],
          "name": "my_func",
          "op": undefined,
          "rhs": undefined,
          "type": "function",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where functions > boolean rhs 1`] = `
{
  "input": "my_func():true",
  "result": {
    "ast": {
      "conditions": [
        {
          "args": [],
          "name": "my_func",
          "op": "",
          "rhs": {
            "parsed": true,
            "type": "boolean",
            "value": "true",
          },
          "type": "function",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where functions > no args 1`] = `
{
  "input": "full_name()",
  "result": {
    "ast": {
      "conditions": [
        {
          "args": [],
          "name": "full_name",
          "op": undefined,
          "rhs": undefined,
          "type": "function",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where functions > no args with space 1`] = `
{
  "input": "full_name(  )",
  "result": {
    "ast": {
      "conditions": [
        {
          "args": [],
          "name": "full_name",
          "op": undefined,
          "rhs": undefined,
          "type": "function",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where functions > number arguments 1`] = `
{
  "input": "my_func(2, 0.5, -2)",
  "result": {
    "ast": {
      "conditions": [
        {
          "args": [
            {
              "position": 0,
              "type": "function_arg",
              "vals": [
                {
                  "parsed": 2,
                  "type": "number",
                  "value": "2",
                },
              ],
            },
            {
              "position": 1,
              "type": "function_arg",
              "vals": [
                {
                  "parsed": 0.5,
                  "type": "number",
                  "value": "0.5",
                },
              ],
            },
            {
              "position": 2,
              "type": "function_arg",
              "vals": [
                {
                  "parsed": -2,
                  "type": "number",
                  "value": "-2",
                },
              ],
            },
          ],
          "name": "my_func",
          "op": undefined,
          "rhs": undefined,
          "type": "function",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where functions > relative date rhs 1`] = `
{
  "input": "last_seen():-7d",
  "result": {
    "ast": {
      "conditions": [
        {
          "args": [],
          "name": "last_seen",
          "op": "",
          "rhs": {
            "parsed": 7,
            "sign": "-",
            "type": "relative_date",
            "unit": "d",
            "value": "7",
          },
          "type": "function",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where functions > single string arg 1`] = `
{
  "input": "count(events)",
  "result": {
    "ast": {
      "conditions": [
        {
          "args": [
            {
              "position": 0,
              "type": "function_arg",
              "vals": [
                {
                  "type": "text",
                  "value": "events",
                },
              ],
            },
          ],
          "name": "count",
          "op": undefined,
          "rhs": undefined,
          "type": "function",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where functions > with filter argument 1`] = `
{
  "input": "count(projects, plan:free !has:email):>5",
  "result": {
    "ast": {
      "conditions": [
        {
          "args": [
            {
              "position": 0,
              "type": "function_arg",
              "vals": [
                {
                  "type": "text",
                  "value": "projects",
                },
              ],
            },
            {
              "position": 1,
              "type": "function_arg",
              "vals": [
                {
                  "lhs": {
                    "type": "qualifier_key",
                    "value": "plan",
                  },
                  "op": "",
                  "rhs": {
                    "type": "text",
                    "value": "free",
                  },
                  "type": "qualifier",
                },
                {
                  "lhs": {
                    "type": "qualifier_key",
                    "value": "has",
                  },
                  "negated": true,
                  "op": "",
                  "rhs": {
                    "type": "text",
                    "value": "email",
                  },
                  "type": "qualifier",
                },
              ],
            },
          ],
          "name": "count",
          "op": ">",
          "rhs": {
            "parsed": 5,
            "type": "number",
            "value": "5",
          },
          "type": "function",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where functions > with quoted argument 1`] = `
{
  "input": "count(projects, "free search"):>0.5",
  "result": {
    "ast": {
      "conditions": [
        {
          "args": [
            {
              "position": 0,
              "type": "function_arg",
              "vals": [
                {
                  "type": "text",
                  "value": "projects",
                },
              ],
            },
            {
              "position": 1,
              "type": "function_arg",
              "vals": [
                {
                  "quoted": true,
                  "type": "text",
                  "value": "free search",
                },
              ],
            },
          ],
          "name": "count",
          "op": ">",
          "rhs": {
            "parsed": 0.5,
            "type": "number",
            "value": "0.5",
          },
          "type": "function",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where functions > with rhs 1`] = `
{
  "input": "divide(projects, max_projects):>0.5",
  "result": {
    "ast": {
      "conditions": [
        {
          "args": [
            {
              "position": 0,
              "type": "function_arg",
              "vals": [
                {
                  "type": "text",
                  "value": "projects",
                },
              ],
            },
            {
              "position": 1,
              "type": "function_arg",
              "vals": [
                {
                  "type": "text",
                  "value": "max_projects",
                },
              ],
            },
          ],
          "name": "divide",
          "op": ">",
          "rhs": {
            "parsed": 0.5,
            "type": "number",
            "value": "0.5",
          },
          "type": "function",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where in filters > negation 1`] = `
{
  "input": "!user:[jane, john]",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "user",
          },
          "negated": true,
          "op": "",
          "rhs": {
            "type": "bracket_list",
            "values": [
              {
                "type": "text",
                "value": "jane",
              },
              {
                "type": "text",
                "value": "john",
              },
            ],
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where in filters > numbers 1`] = `
{
  "input": "release:[12.0]",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "release",
          },
          "op": "",
          "rhs": {
            "type": "bracket_list",
            "values": [
              {
                "parsed": 12,
                "type": "number",
                "value": "12.0",
              },
            ],
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where in filters > quoted 1`] = `
{
  "input": "release:[jane, "john doe"]",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "release",
          },
          "op": "",
          "rhs": {
            "type": "bracket_list",
            "values": [
              {
                "type": "text",
                "value": "jane",
              },
              {
                "quoted": true,
                "type": "text",
                "value": "john doe",
              },
            ],
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where in filters > works 1`] = `
{
  "input": "user:[jane, john]",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "user",
          },
          "op": "",
          "rhs": {
            "type": "bracket_list",
            "values": [
              {
                "type": "text",
                "value": "jane",
              },
              {
                "type": "text",
                "value": "john",
              },
            ],
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where operators > < 1`] = `
{
  "input": "num_members:<5.0",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "num_members",
          },
          "op": "<",
          "rhs": {
            "parsed": 5,
            "type": "number",
            "value": "5.0",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where operators > <= 1`] = `
{
  "input": "num_members:<=5.0",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "num_members",
          },
          "op": "<=",
          "rhs": {
            "parsed": 5,
            "type": "number",
            "value": "5.0",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where operators > = 1`] = `
{
  "input": "num_members:=5",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "num_members",
          },
          "op": "=",
          "rhs": {
            "parsed": 5,
            "type": "number",
            "value": "5",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where operators > > 1`] = `
{
  "input": "num_members:>5.0",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "num_members",
          },
          "op": ">",
          "rhs": {
            "parsed": 5,
            "type": "number",
            "value": "5.0",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where operators > >= 1`] = `
{
  "input": "num_members:>=5",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "num_members",
          },
          "op": ">=",
          "rhs": {
            "parsed": 5,
            "type": "number",
            "value": "5",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where relative date > + 1`] = `
{
  "input": "renews_at:+30d",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "renews_at",
          },
          "op": "",
          "rhs": {
            "parsed": 30,
            "sign": "+",
            "type": "relative_date",
            "unit": "d",
            "value": "30",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where relative date > - 1`] = `
{
  "input": "last_seen:-1d",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "last_seen",
          },
          "op": "",
          "rhs": {
            "parsed": 1,
            "sign": "-",
            "type": "relative_date",
            "unit": "d",
            "value": "1",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where relative date > floats are not allowed 1`] = `
{
  "input": "renews_at:+1.2s",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "renews_at",
          },
          "op": "",
          "rhs": {
            "type": "text",
            "value": "",
          },
          "type": "qualifier",
        },
        {
          "args": [],
          "name": "s",
          "op": undefined,
          "rhs": undefined,
          "sort": "asc",
          "type": "function",
        },
      ],
      "type": "where_expr",
    },
    "errors": [
      [NoViableAltException: Expecting: one of these possible Token sequences:
  1. [LQuote]
  2. [Number]
  3. [Boolean]
  4. [RelativeDate]
  5. [Iso8601Date]
  6. [Identifier]
but found: '+'],
      [MismatchedTokenException: Expecting token of type --> Identifier <-- but found --> '1.2' <--],
      [MismatchedTokenException: Expecting token of type --> LParen <-- but found --> '' <--],
    ],
  },
}
`;

exports[`where relative date > with other filters 1`] = `
{
  "input": "plan:free renews_at:+4w",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "plan",
          },
          "op": "",
          "rhs": {
            "type": "text",
            "value": "free",
          },
          "type": "qualifier",
        },
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "renews_at",
          },
          "op": "",
          "rhs": {
            "parsed": 4,
            "sign": "+",
            "type": "relative_date",
            "unit": "w",
            "value": "4",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where text filters > .-_[0-9] in keys 1`] = `
{
  "input": "payments.plan-type_is2:free",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "payments.plan-type_is2",
          },
          "op": "",
          "rhs": {
            "type": "text",
            "value": "free",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where text filters > @.-_ in values 1`] = `
{
  "input": "email:jane@example.com type:public-company_x",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "email",
          },
          "op": "",
          "rhs": {
            "type": "text",
            "value": "jane@example.com",
          },
          "type": "qualifier",
        },
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "type",
          },
          "op": "",
          "rhs": {
            "type": "text",
            "value": "public-company_x",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where text filters > keys that start with a $ 1`] = `
{
  "input": "$last_seen:foo",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "$last_seen",
          },
          "op": "",
          "rhs": {
            "type": "text",
            "value": "foo",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where text filters > keys with a period and $ 1`] = `
{
  "input": "company.$id:123",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "company.$id",
          },
          "op": "",
          "rhs": {
            "parsed": 123,
            "type": "number",
            "value": "123",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where text filters > multiple 1`] = `
{
  "input": "plan:free age:5",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "plan",
          },
          "op": "",
          "rhs": {
            "type": "text",
            "value": "free",
          },
          "type": "qualifier",
        },
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "age",
          },
          "op": "",
          "rhs": {
            "parsed": 5,
            "type": "number",
            "value": "5",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where text filters > negation 1`] = `
{
  "input": "!plan:free",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "plan",
          },
          "negated": true,
          "op": "",
          "rhs": {
            "type": "text",
            "value": "free",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where text filters > negation in quoted 1`] = `
{
  "input": "!user:"john doe"",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "user",
          },
          "negated": true,
          "op": "",
          "rhs": {
            "quoted": true,
            "type": "text",
            "value": "john doe",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where text filters > partial filter still parses 1`] = `
{
  "input": "age:> plan:free user:",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "age",
          },
          "op": ">",
          "rhs": undefined,
          "type": "qualifier",
        },
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "plan",
          },
          "op": "",
          "rhs": {
            "type": "text",
            "value": "free",
          },
          "type": "qualifier",
        },
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "user",
          },
          "op": undefined,
          "rhs": undefined,
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [
      [NoViableAltException: Expecting: one of these possible Token sequences:
  1. [LBracket]
  2. [+]
  3. [Minus]
  4. [LQuote]
  5. [Number]
  6. [Boolean]
  7. [RelativeDate]
  8. [Iso8601Date]
  9. [Identifier]
but found: 'plan:'],
    ],
  },
}
`;

exports[`where text filters > quoted value 1`] = `
{
  "input": "user:"john doe"",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "user",
          },
          "op": "",
          "rhs": {
            "quoted": true,
            "type": "text",
            "value": "john doe",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;

exports[`where text filters > works 1`] = `
{
  "input": "plan:free",
  "result": {
    "ast": {
      "conditions": [
        {
          "lhs": {
            "type": "qualifier_key",
            "value": "plan",
          },
          "op": "",
          "rhs": {
            "type": "text",
            "value": "free",
          },
          "type": "qualifier",
        },
      ],
      "type": "where_expr",
    },
    "errors": [],
  },
}
`;
